pkg/util/dynclient/dynclient.go

```diff
@@ -0,0 +1,108 @@
+package dynclient
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	dynamic "k8s.io/client-go/deprecated-dynamic"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+)
+
+type DynClient interface {
+	GetOrList(ctx context.Context, kind, namespace, name string) (*unstructured.UnstructuredList, error)
+}
+
+type dynclient struct {
+	log *logrus.Entry
+
+	cli *discovery.DiscoveryClient
+	dyn dynamic.ClientPool
+	grs []*restmapper.APIGroupResources
+}
+
+func New(log *logrus.Entry, restconfig *rest.Config) (DynClient, error) {
+	d := &dynclient{
+		log: log,
+	}
+
+	var err error
+	d.cli, err = discovery.NewDiscoveryClientForConfig(restconfig)
+	if err != nil {
+		return nil, err
+	}
+
+	grs, err := restmapper.GetAPIGroupResources(d.cli)
+	if err != nil {
+		return nil, err
+	}
+	d.grs = grs
+
+	rm := restmapper.NewDiscoveryRESTMapper(d.grs)
+	d.dyn = dynamic.NewClientPool(restconfig, rm, dynamic.LegacyAPIPathResolverFunc)
+
+	return d, nil
+}
+
+// contains returns true if haystack contains needle
+func contains(haystack []string, needle string) bool {
+	for _, s := range haystack {
+		if s == needle {
+			return true
+		}
+	}
+	return false
+}
+
+func (d *dynclient) GetOrList(ctx context.Context, kind, namespace, name string) (*unstructured.UnstructuredList, error) {
```

Comment:
We should be able to handle a slightly underspecified groupkind here, so the following inputs should all work:
"Pod", "ClusterVersion.config.openshift.io", "clusterversion"

For now, let's not enable the end user to specify the version, let's just use the server preferred version.